### PR TITLE
replaced downsampler with wasm version

### PR DESCRIPTION
**Pull request by mrrostam**
- refactored the downsampler_worker and replaced the js version with the wasm one.

### CODE REVIEW

File: src/web_voice_processor.ts [hunk 1]
```diff
@@ -122,6 +157,10 @@ export class WebVoiceProcessor {
           this._audioDumpReject = null;
           break;
         }
+        default: {
+          console.assert(`unexpected recieved command: ${event.data.command}`);
```
Comment (dbartle) [inline src/web_voice_processor.ts]: recieved => received

Received unexpected command

use `console.warn`

File: src/downsampling_worker.ts [hunk 2]
```diff
@@ -9,83 +9,322 @@
...
-import { DownsamplingWorkerRequest } from './worker_types';
+import {
+  DownsamplingWorkerRequest,
+  DownsamplerWasmOutput,
+  DownsamplerInterface,
+} from './worker_types';
+import { wasiSnapshotPreview1Emulator } from './wasi_snapshot';
+import { WASM_BASE64 } from './downsampler_base64';
+
+class Downsampler implements DownsamplerInterface {
```
Comment (dbartle) [inline src/downsampling_worker.ts]: move Downsampler class to separate `downsampler.ts` file

File: src/worker_types.ts [hunk 3]
```diff
@@ -48,8 +52,29 @@ export type DownsamplingWorkerRequest =
...
+  pvDownsamplerReset: CallableFunction,
+  pvDownsamplerDelete: CallableFunction,
+}
+
+export interface DownsamplerInterface {
+  delete(): void;
+  getrequiredNumInputSamples(numSample: number): number;
+  process(inputBuffer: Int16Array, inputBufferSize: number, outputBuffer: Int16Array): number;
+  reset(): void;
+}
```
Comment (dbartle) [inline src/worker_types.ts]: move these types to `downsampler.ts`

File: src/web_voice_processor.ts [hunk 4]
```diff
@@ -56,11 +57,56 @@ export class WebVoiceProcessor {
       audio: true,
     });
 
-    return new WebVoiceProcessor(microphoneStream, options);
+    const audioContext = new (window.AudioContext ||
+      // @ts-ignore window.webkitAudioContext
+      window.webkitAudioContext)();
+    const audioSource = audioContext.createMediaStreamSource(microphoneStream);
+
+    const downsamplingWorker = new DsWorker() as DownsamplingWorker;
```
Comment (dbartle) [inline src/web_voice_processor.ts]: Use the ___WorkerFactory pattern from ppn/rhn/pv binding to abstract away this (including `new`)

File: src/web_voice_processor.ts [hunk 5]
```diff
@@ -56,11 +57,56 @@ export class WebVoiceProcessor {
...
+      microphoneStream,
+      audioContext,
+      audioSource,
+      downsamplingWorker,
+      options,
+    );
   }
 
-  constructor(
+  private static async initDsWorker(
```
Comment (dbartle) [inline src/web_voice_processor.ts]: move to DownsamplerWorkerFactory

File: src/downsampling_worker.ts [hunk 6]
```diff
@@ -9,83 +9,322 @@
...
+
+  public reset(): void {
+    this._pvDownsamplerReset(this._objectAddress);
+  }
+
+  public delete(): void {
+    this._pvDownsamplerDelete(this._objectAddress);
+  }
+
+  public getrequiredNumInputSamples(numSample: number): number {
```
Comment (dbartle) [inline src/downsampling_worker.ts]: getNumRequiuredInputSamples

File: src/downsampling_worker.ts [hunk 7]
```diff
@@ -9,83 +9,322 @@
...
+    }
+    return processedSamples;
+  }
+
+  public reset(): void {
+    this._pvDownsamplerReset(this._objectAddress);
+  }
+
+  public delete(): void {
+    this._pvDownsamplerDelete(this._objectAddress);
```
Comment (dbartle) [inline src/downsampling_worker.ts]: what happens if we call this twice?

File: src/downsampling_worker.ts [hunk 8]
```diff
@@ -9,83 +9,322 @@
...
+  private static arrayBufferToStringAtIndex(
+    arrayBuffer: Uint8Array,
+    index: number,
+  ): string {
+    let stringBuffer = '';
+    let indexBuffer = index;
+    while (true) {
+      if (arrayBuffer[indexBuffer] === 0) {
+        break;
+      }
```
Comment (dbartle) [inline src/downsampling_worker.ts]: I think we can just put the `if` condition inside of `while`

File: src/downsampling_worker.ts [hunk 9]
```diff
@@ -9,83 +9,322 @@
...
-  _audioDumpNumFrames = durationMs * (PV_FRAME_LENGTH / PV_SAMPLE_RATE);
+  _audioDumpNumSamples = (durationMs * PV_SAMPLE_RATE) / 1000;
   _audioDumpActive = true;
   _audioDumpBufferIndex = 0;
-  _audioDumpBuffer = new Int16Array(_audioDumpNumFrames * _frameLength);
+  _audioDumpBuffer = new Int16Array(_audioDumpNumSamples);
 }
 
 function processAudio(inputFrame: Float32Array): void {
+  console.assert(inputFrame.constructor === Float32Array);
```
Comment (dbartle) [inline src/downsampling_worker.ts]: good runtime type check here. go ahead and throw an Error if the condition is not met.

File: src/downsampling_worker.ts [hunk 10]
```diff
@@ -9,83 +9,322 @@
...
-      ) {
-        sum += _inputBuffer[inputIndex];
-        num++;
-        inputIndex++;
-      }
-      outputFrame[outputIndex] = sum / num;
-      outputIndex++;
-    }
+  while (_inputBuffer.length > 0) {
+    const NumInputSamples =
```
Comment (dbartle) [inline src/downsampling_worker.ts]: numInputSamples

File: src/downsampling_worker.ts [hunk 11]
```diff
@@ -9,83 +9,322 @@
...
+      _downsampler.getrequiredNumInputSamples(_outputframeLength);
+    const NumInputSamplesToRead = Math.min(
```
Comment (dbartle) [inline src/downsampling_worker.ts]: numInputSamplesToRead

**Review by dbartle — CHANGES_REQUESTED** [review]

File: src/downsampling_worker.ts (hunk 7 above)
Comment (mrrostam) [reply src/downsampling_worker.ts]: nothing. it's checked in C whether the object is deleted or not.

File: src/downsampling_worker.ts [hunk 12]
```diff
@@ -11,81 +11,94 @@
...
+  _audioDumpNumSamples = (durationMs * PV_SAMPLE_RATE) / 1000;
   _audioDumpActive = true;
   _audioDumpBufferIndex = 0;
-  _audioDumpBuffer = new Int16Array(_audioDumpNumFrames * _frameLength);
+  _audioDumpBuffer = new Int16Array(_audioDumpNumSamples);
 }
 
 function processAudio(inputFrame: Float32Array): void {
+  if (inputFrame.constructor !== Float32Array) {
+    throw new Error(`Invalid input frame type :${inputFrame.constructor}.`);
```
Comment (dbartle) [inline src/downsampling_worker.ts]: `Invalid inputFrame type: ${typeof inputFrame}. Expected Float32Array.`

File: src/downsampling_worker.ts [hunk 13]
```diff
@@ -11,81 +11,94 @@
...
-
-  console.assert(Number.isInteger(_inputSampleRate));
-  console.assert(Number.isInteger(_outputSampleRate));
-  console.assert(Number.isInteger(_frameLength));
-
-  _inputBuffer = [];
+): Promise<void> {
+  console.assert(Number.isInteger(inputSampleRate));
+  console.assert(Number.isInteger(outputSampleRate));
+  console.assert(Number.isInteger(frameLength));
```
Comment (dbartle) [inline src/downsampling_worker.ts]: replace these assertions with throwing errors instead

File: src/worker_types.ts [hunk 14]
```diff
@@ -48,8 +52,20 @@ export type DownsamplingWorkerRequest =
...
+
+export interface DownsamplerInterface {
+  delete(): void;
+  getNumRequiredInputSamples(numSample: number): number;
+  process(
+    inputBuffer: Int16Array,
+    inputBufferSize: number,
+    outputBuffer: Int16Array,
+  ): number;
+  reset(): void;
```
Comment (dbartle) [inline src/worker_types.ts]: Add comments to each member of the interface with `/** */`

File: src/downsampling_worker.ts [hunk 15]
```diff
@@ -11,81 +11,94 @@
...
+    const processedSamples = _downsampler.process(
+      _inputBuffer.slice(0, numInputSamplesToRead),
+      numInputSamplesToRead,
+      _outputBuffer,
+    );
 
     if (_audioDumpActive) {
-      _audioDumpBuffer.set(outputFrame, _audioDumpBufferIndex * _frameLength);
-      _audioDumpBufferIndex++;
+      const NumSamplesToCopy = Math.min(
```
Comment (dbartle) [inline src/downsampling_worker.ts]: numSamplesToCopy

File: src/downsampler.ts [hunk 16]
```diff
@@ -0,0 +1,239 @@
...
+  }
+
+  public getNumRequiredInputSamples(numSample: number): number {
+    return this._pvDownsamplerConvertNumSamplesToInputSampleRate(
+      this._objectAddress,
+      numSample,
+    );
+  }
+
+  private static arrayBufferToStringAtIndex(
```
Comment (dbartle) [inline src/downsampler.ts]: this method and the one below don't need to be part of the class; you can move them outside. they can live in this file, or they could be put in a `utils.ts` file and imported. Either is acceptable.

File: src/downsampler.ts [hunk 17]
```diff
@@ -0,0 +1,239 @@
...
+  ): string {
+    let stringBuffer = '';
+    let indexBuffer = index;
+    while (arrayBuffer[indexBuffer] === 0) {
+      stringBuffer += String.fromCharCode(arrayBuffer[indexBuffer++]);
+    }
+    return stringBuffer;
+  }
+
+  private static base64ToUint8Array(base64String: string): Uint8Array {
```
Comment (dbartle) [inline src/downsampler.ts]: move outside of class

**Review by dbartle — CHANGES_REQUESTED** [review]

File: src/downsampler.ts (hunk 17 above)
Comment (mrrostam) [reply src/downsampler.ts]: all comments are resolved as requested.

**Pull request merged**